### PR TITLE
Introduce Project.HasCompleteReferencesAsync

### PR DESCRIPTION
**Pull request by jasonmalinowski**
When you get a compilation from a project in a workspace, it's possible that the compilation doesn't have all the references you expect. If the project has references to other languages, we have to produce reference assemblies, a process that might fail. If it fails, your compilation just doesn't get references you thought it'd get. Now with this API, you can ask if the compilation is going to have such missing references, and act accordingly.

**NB:** This pull request also includes a number of other small teaks that I noticed while trying to understand how `CompilationTracker` works. The commits are broken up, so feel free to look at them and understand which is which. I'm happy to discuss those as well and elide any changes people don't like.

**Still Coming:** An API to explicitly state that projects are missing references outside the purview of the workspaces API.

**Reviewers:** @mattwar, @heejaechang, @mavasani, @dotnet/roslyn-ide

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
Any chance I can get a code review, @dotnet/roslyn-ide, @mattwar, @heejaechang? I just made the API internal so it's not screwing up any of our API reviews that are currently in progress. We can make it public once we've got a consumer to vet it.

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs [hunk 1]
```diff
@@ -390,6 +390,15 @@ public Task<Compilation> GetCompilationAsync(CancellationToken cancellationToken
         }
 
         /// <summary>
+        /// Determines if the compilation returned by <see cref="GetCompilationAsync"/> has all the references it's expected to have.
+        /// </summary>
+        // TODO: make this public
+        internal Task<bool> GetCompilationHasIncompleteReferencesAsync(CancellationToken cancellationToken = default(CancellationToken))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Project.cs]: HasIncomplete feels very weird to me.  it's like having an "IsNotXXX" method/property.  Perhaps it should be "HasCompleteReferences"?

**Comment by CyrusNajmabadi** [comment]
Who is the expected consumer for this?  

My biggest issue seems to be with the name.  It seems overwrought (for lack of a better term).  "Incomplete References" just isn't a term that feels right to me.  Maybe something like "CompilationContainsAllProjectReferences" or something like that might be better.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.State.cs [hunk 2]
```diff
@@ -15,45 +15,50 @@ public partial class Solution
...
+                /// The best compilation that is available.  May be an in-progress, full declaration, or a final compilation.
+                /// </summary>
                 public ValueSource<Compilation> Compilation { get; }
 
-                // The Final compilation if available, otherwise an empty IValueSource
+                /// <summary>
+                /// Specifies if there are references that got dropped in the production of <see cref="FinalCompilation"/>. This can return
+                /// null if the state isn't at the point where it would know, and it's necessary to transition to <see cref="FinalState"/> to figure that out.
+                /// </summary>
+                public virtual bool? HasIncompleteReferences => null;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.State.cs]: This feels *weird*.  You'd only be ever able to ask "does my compilation have incomplete references" if you've actually tried to get the 'final compilation'.  As such, I don't think this book should be here.  It should only be on FinalState.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs [hunk 3]
```diff
@@ -307,7 +308,23 @@ public Task<Compilation> GetCompilationAsync(Solution solution, CancellationToke
...
+                    return GetOrBuildCompilationInfoAsync(solution, lockGate: true, cancellationToken: cancellationToken)
+                        .ContinueWith(t => t.Result.Compilation, cancellationToken, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+                }
+            }
+
+            public Task<bool> GetCompilationHasIncompleteReferencesAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                var state = this.ReadState();
+
+                if (state.HasIncompleteReferences.HasValue)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs]: This is *very* weird.  Say we've only computed up to the 'in progress' state.  From your code elsewhere, you initialize this state with "true" for "HasIncompleteReferences".

But this doesn't seem accurate to me.  You're effectively saying "the compilation is missing references!".  But if I actually went and then asked for the compilation it might have all references.  

Furthermore, it now seems like the Project might return different results for this method depending on when it was called.  i.e. if some other thread took us to the "final state" or not.  That makes this API seem observably mutable, and I'm really wary about having that in the API.

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Project.cs]: For some reason today that sounds better than I think when I last tried it. It'll be HasCompleteReferencesAsync() since it's an async method. I guess the thing that bugs me most here is I want async properties or something...

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.State.cs (hunk 2 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.State.cs]: This is how this pattern is done in the compilation tracker. If you call `GetCompilation()` it first checks the state and sees if it already has a return value for `FinalCompilation` below, and returns it. Otherwise it goes through it's process to bring it to the final state. So this is _consistent_ with the pattern. I agree it's somewhat strange though, but we could try moving both of them down.

Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.State.cs]: There might also be some cases where we know references are missing even earlier in the process, and there's no reason to have to force compilation generation for those cases....so I'm not morally against this particular one.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi: The expected consumer of this is our diagnostics engine, to avoid creating massive numbers of errors in the error list when we know something is wrong. Right now if a skeleton reference fails to build, the diagnostic engine has no practical way to know that reference is missing, and so we end up creating tens of thousands of errors in the error list about missing types. The hope is we will consume this API and simply not show those errors in the error list. (The exact question of what will and won't be shown where is still being refined.)

The reason the name is vague versus something more tuned to project references is we will also allow this to return true in other cases. My plan is to add a workspace hosting API where the host can explicitly state "the references I gave you are incomplete". This we would set from the project system for scenarios like a NuGet package restore failing, a design time build failed, etc. Basically, scenarios where we _know_ things are going to be screwed up that there's just no value in giving the user a million errors in the error list that hide the root cause.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs (hunk 3 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs]: Where do I initialize the state with 'true'? InProgress state will return null, thus forcing the compilation to be finalized. The only override (today) that could return something other than null is in FinalState.

The intent of the API is for this to be as side-effect free as GetCompilation is, and calling them in either order shouldn't cause any effects. I _believe_ I got this right, but I confess that it took me a half of a day to understand how the code worked...

**Comment by mattwar** [comment]
Why is the API internal?

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTranslationAction.cs [hunk 4]
```diff
@@ -50,20 +50,6 @@ public static CompilationTranslationAction TouchDocument(DocumentState oldState,
             }
 
             #endregion
-
-            public static void CheckKnownActions(CompilationTranslationAction translate)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTranslationAction.cs]: Why was this removed?

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs [hunk 5]
```diff
@@ -1754,9 +1754,6 @@ private Solution WithTextDocumentState(TextDocumentState newDocument, bool textC
             ImmutableDictionary<string, ImmutableArray<DocumentId>> newLinkedFilesMap = null,
             bool forkTracker = true)
         {
-            // make sure we are getting only known translate actions
-            CompilationTranslationAction.CheckKnownActions(translate);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: And this?

**Comment by CyrusNajmabadi** [comment]
Jason, LGTM (except for the pieces of code that were removed).

**Comment by Pilchie** [comment]
LGTM

**Comment by jasonmalinowski** [comment]
@mattwar: since we're sufficiently close to the update, I don't want to make it public until we have some minimum proof it actually is useful.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi: I removed those bits because they had no value. The CheckKnownActions just explicitly listed each and every derived type of the base type, and asserted that the instance was one of those. All it would do is (a) cause is to do runtime checks because why not and (b) ensure that anybody adding or removing translation actions gets to update more code for kicks.

**Comment by CyrusNajmabadi** [comment]
LGTM then!  :+1:

**Comment by heejaechang** [comment]
I added the CheckKnownAction check for this in the ForkProject

var modifiedDocumentOnly = translate is CompilationTranslationAction.TouchDocumentAction;
            var newLatestProjectVersion = modifiedDocumentOnly ? _lazyLatestProjectVersion : new Lazy<VersionStamp>(() => newProjectState.Version);

worrying that someone adds derived type that would break this. 

probably adding a method "IsProjectVersionChanged" or something in translation action could be better.

**Comment by jasonmalinowski** [comment]
@heejaechang: should we just seal the type?

**Comment by heejaechang** [comment]
@jasonmalinowski  even if we seal all existing derived type, one can add new derived type from base one.

**Comment by jasonmalinowski** [comment]
@heejaechang: so? If somebody does that, they'd just update that list without thinking about it. It was asserting *all* derived types, nothing more precise.

**Comment by jasonmalinowski** [comment]
test mac please

**Comment by heejaechang** [comment]
@jasonmalinowski hmmm how about we put that logic in the transition then, like transition.GetProjectVersion(...)

**Comment by jasonmalinowski** [comment]
@heejaechang: I still don't get what the logic was. The code asserted it was a certain derived type, when the derived type was "all of them".

**Comment by heejaechang** [comment]
its intension was, if someone introduced a new derived type from the base derived type (yes, I should have sealed existing derived type so that people can’t derived from existing derived types), we crash so that people who added the derived type can question why the check is there.

What I am saying, putting the logic (getting new project version) in transition itself is probably better approach.

**Comment by jasonmalinowski** [comment]
@heejaechang: the check didn't explain why it was there...so it just looked like dead code. :smile:

**Comment by heejaechang** [comment]
ya, I should have done that :)

**Comment by heejaechang** [comment]
@jasonmalinowski by the way, I didn't write the comment to block your checkin, I wrote it just to let you know why the check was there.

**Comment by jasonmalinowski** [comment]
@heejaechang: I didn't take it as a block...was hoping to make Jenkins a bit happier first.

**Comment by jasonmalinowski** [comment]
retest this please

**Comment by jasonmalinowski** [comment]
retest linux please
